[GTK][WPE] Add API to replace and retrieve the entire cookie jar

pgorszkowski-igalia opened the pull request:
#### 3f2a8e4ebeb1b855e6080f149fc234e5beae567b
<pre>
[GTK][WPE] Add API to replace and retrieve the entire cookie jar
<a href="https://bugs.webkit.org/show_bug.cgi?id=254987">https://bugs.webkit.org/show_bug.cgi?id=254987</a>

Reviewed by Carlos Garcia Campos, Adrian Perez de Castro and Michael Catanzaro.

Add possibility to replace and get all cookies, from each domain via new API.

* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
(WebCore::NetworkStorageSession::setCookieJar):
* Source/WebKit/NetworkProcess/Cookies/WebCookieManager.cpp:
(WebKit::WebCookieManager::setCookieJar):
* Source/WebKit/NetworkProcess/Cookies/WebCookieManager.h:
* Source/WebKit/NetworkProcess/Cookies/WebCookieManager.messages.in:
* Source/WebKit/UIProcess/API/APIHTTPCookieStore.cpp:
(API::HTTPCookieStore::setCookieJar):
(API::HTTPCookieStore::getCookieJar):
* Source/WebKit/UIProcess/API/APIHTTPCookieStore.h:
* Source/WebKit/UIProcess/API/glib/WebKitCookieManager.cpp:
(webkit_cookie_manager_set_cookie_jar):
(webkit_cookie_manager_set_cookie_jar_finish):
(webkit_cookie_manager_get_cookie_jar):
(webkit_cookie_manager_get_cookie_jar_finish):
* Source/WebKit/UIProcess/API/glib/WebKitCookieManager.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestCookieManager.cpp:
(testCookieManagerSetGetCookieJar):
(beforeAll):

Canonical link: <a href="https://commits.webkit.org/265834@main">https://commits.webkit.org/265834@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a21c4f44f2048c375477ae1f94e7ac2feb14d4ca

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7539 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7806 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7989 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9178 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7728 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7550 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/9879 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7731 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10610 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7670 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/8888 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6963 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9286 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6196 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6863 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14569 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7360 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6985 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10330 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7475 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6108 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6816 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2941 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11024 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7209 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7834d803d59265b80de072763f107bb82213fe90. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12639

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/023890af0792cdd92adca37364f67cfd78c7d9a3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12639

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4280e043ae0ffb75aaa8ade63fe30460ad69b554. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12639

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/56ae57ef74ddd2d660a9762fc4aedc3da1dc383e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12639

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a22b0144273cc921a15c6bdbc8b9866b4524343e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12639

aperezdc commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -410,6 +410,30 @@ void NetworkStorageSession::setCookie(const Cookie& cookie)
+    // Delete existent cookies and add the new ones. During the process, disable
+    // signals from the cookie jar so we don't get a ton of them.
+    guint signalId = g_signal_lookup("changed", soup_cookie_jar_get_type());
+    GArray* blockedHandlerIdArray = g_array_new(FALSE, FALSE, sizeof(gulong));
```
I would use `WTF::Vector<unsigned long>` instead of `GArray`.

aperezdc commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -410,6 +410,30 @@ void NetworkStorageSession::setCookie(const Cookie& cookie)
+    gulong handlerId;
+    while ((handlerId = g_signal_handler_find(jar, (GSignalMatchType)(G_SIGNAL_MATCH_ID | G_SIGNAL_MATCH_UNBLOCKED), signalId, 0, NULL, NULL, NULL))) {
+        g_array_append_val(blockedHandlerIdArray, handlerId);
+        g_signal_handler_block(jar, handlerId);
```
Even better, why not using `g_signal_handlers_block_matched()`, and then `g_signal_handlers_unblock_matched()`?

pgorszkowski-igalia replied:
`g_signal_handlers_block_matched` does not support G_SIGNAL_MATCH_ID (base on test and documentation: https://docs.gtk.org/gobject/func.signal_handlers_block_matched.html

But I found something similar which can be used and it works:
```
    guint signalId = g_signal_lookup("changed", soup_cookie_jar_get_type());
    gulong handler = g_signal_handler_find (jar, G_SIGNAL_MATCH_ID, signalId, 0, NULL, NULL, NULL);

    g_signal_handler_block(jar, handler);
```

I created also a patch to glib which adds a possibility to use  `g_signal_handlers_block_matched` with G_SIGNAL_MATCH_ID:
https://github.com/pgorszkowski-igalia/glib/commit/0eee1123840003f611cec484e5484ff6cf520e3d

I will try to contribute with this patch to glib project

pgorszkowski-igalia replied:
And one more thing which should be added here (after discussion with @aperezdc) : at the end of the `NetworkStorageSession::setCookieJar` one signal  "changed" should be sent.

pgorszkowski-igalia replied:
MR for glib change: https://gitlab.gnome.org/GNOME/glib/-/merge_requests/3376

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitCookieManager.cpp:
```diff
@@ -517,3 +517,105 @@ void webkit_cookie_manager_delete_all_cookies(WebKitCookieManager* manager)
+/**
+ * webkit_cookie_manager_set_cookie_jar:
+ * @cookie_manager: a #WebKitCookieManager
+ * @cookies: (element-type SoupCookie) (transfer full): A #GList of #SoupCookie to be added
```
Note that the `WebCore::Cookie` constructor only reads fields from the `SoupCookie`, making copies of values. Therefore this should be `(transfer none)` because the function is _not_ taking ownership of (and not freeing, nor consuming) the passed list of cookies.

aperezdc reviewed (changes requested):
I left a couple of comments. My main doubt: Is there some use case of an application that needs this new API? If yes, mentioning an example use case would help, otherwise we tend to be reluctant to add new API “just because sounds nice to have”.

magomez commented:
> I left a couple of comments. My main doubt: Is there some use case of an application that needs this new API? If yes, mentioning an example use case would help, otherwise we tend to be reluctant to add new API “just because sounds nice to have”.

The main use case here is backup and restore the entire cookie jar. I guess this is useful when migrating user profiles or maybe to ensure that there isn't data loss when when upgrading versions.

aperezdc commented:
Today I was giving this a second thought, and I have a couple more of comments with ideas. Of course these are open for dsicussion discussion and it would be great to know the opinions of other reviewers (specially @carlosgcampos and/or @mcatanzaro):

1. I think we should remove the word “jar” from the method names for consistency: the rest of the method do not have it, and because we are operating on a `WebKitCookieManager` it's already clear that we are operating on some kind of container that handles cookie storage. This means:
  - Using `webkit_cookie_manager_get_all_cookies` better than `webkit_cookie_manager_get_cookie_jar`.
    - Alternatively, allow passing `NULL` as the URL for `webkit_cookie_manager_get_cookies` to skip filtering cookies and returning all instead. I am not super convinced this is a great idea because it involves changing the semantics of existing API... If we wanted this, should have been done earlier before considering the 2022 API stable. I guess we lot an opportunity here.
  - Using `webkit_cookie_manager_set_all_cookies` better than `webkit_cookie_manager_set_cookie_jar`.

2. Sometimes programs may want to add multiple cookies without removing the existing ones. I think it would be valuable to add a flag to the `webkit_cookie_manager_set_all_cookies` to let it known whether the passed list replaces or adds cookies. For example Cog can add multiple cookies from the command line, and currently it does it [one by one](https://github.com/Igalia/cog/blob/48347f3a36e6dba75bcfd4f9443730861adec5b0/launcher/cog-launcher.c#L752), and it could instead prepare a list to add all the cookies in one single function call. In this case, we may want to have somethbning like this:

```c
typedef enum {
     WEBKIT_COOKIE_MANAGER_ADD,
     WEBKIT_COOKIE_MANAGER_REPLACE,
     // Maybe others?
} WebKitCookieManagerOperation;

void webkit_cookie_manager_add_cookies(WebKitCookieManager         *manager,
                                       GList                       *cookies,
                                       WebKitCookieManagerOperation operation,
                                       GCancellable                *cancellable,
                                       GAsyncReadyCallback          callback,
                                       gpointer                     userdata);
````

Of course, if people like the idea of allowing more kinds of batch operations, maybe we need a better name, like `webkit_cookie_manager_batch_modify` + `webkit_cookie_manager_batch_modify_finish` 🤔

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/96444a1a0c4c8759d1f211821818658c9e04ad94. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12639

aperezdc commented:
> > The main use case here is backup and restore the entire cookie jar. I guess this is useful when migrating user profiles or maybe to ensure that there isn't data loss when when upgrading versions.
> 
> Well the API looks OK to add regardless -- I can't think of any good reason _not_ to offer this functionality on WebKitCookieManager -- but the use case feels pretty questionable. If you want to back up a user profile and only back up the cookie jar, then you notably miss IndexedDB and localstorage, not to mention all other forms of website data. I don't think using this API for backups would be advisable.

Yes, I think it's good to have the new API added, but I was curious about how people would use it, to try and understand why such functionality has not been requested before&mdash;that's all.

Any thoughts on my suggested naming changes for the methods, and/or adding a flag that would allow appending or replacing to the list of cookies instead of always replacing?

> > Alternatively, allow passing NULL as the URL for webkit_cookie_manager_get_cookies to skip filtering cookies and returning all instead.
> 
> That only works fine if you're using C. Making a parameter nullable is an introspection API break.

That's why I already wrote that it doesn't sound like a good idea and that we had a missed opportunity before we deemed the 2022 GLib API as stable 😉

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/556e9aa92022af27561c8e96993bd837b302b02f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12639

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c1c37580023e041d91340c26dc6d70e766875036. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12639

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/83a974ac1d5323dd3166ebce20fb2fd5fe801c0c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12639

aperezdc reviewed (approved):
I like how this is looking after the latest changes to the public API. Thanks for the update!

We only need a second approval and this would be good to go... Maybe @mcatanzaro can take a look again, as he has already written comments a few days ago.

mcatanzaro commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -410,6 +410,26 @@ void NetworkStorageSession::setCookie(const Cookie& cookie)
+{
+    SoupCookieJar* jar = cookieStorage();
+
+    // Delete existent cookies and add the new ones. During the process, disable
```
```suggestion
    // Delete existing cookies and add the new ones. During the process, disable
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitCookieManager.cpp:
```diff
@@ -517,3 +517,106 @@ void webkit_cookie_manager_delete_all_cookies(WebKitCookieManager* manager)
+ * @cookies: (element-type SoupCookie) (transfer none): a #GList of #SoupCookie to be added
+ * @operation: a #WebKitCookieManagerOperation
+ * @cancellable: (allow-none): a #GCancellable or %NULL to ignore
+ * @callback: (scope async): a #GAsyncReadyCallback to call when the request is satisfied
+ * @user_data: (closure): the data to pass to callback function
```
```suggestion
 * @callback: (scope async): (closure): a #GAsyncReadyCallback to call when the request is satisfied
 * @user_data: the data to pass to callback function
```

The `(closure)` annotation is in the wrong place, a common mistake. If you remember where you copied this from, we should fix that place too (in a separate commit).

mcatanzaro replied:
Actually, don't worry about wherever you copied this from. We need to do a general audit because it's all over the place. I'll add it to my to-do list.

mcatanzaro replied:
Um, I amended my suggestion. This could be:

```
 * @callback: (scope async): (closure user_data): a #GAsyncReadyCallback to call when the request is satisfied
 * @user_data: the data to pass to callback function
```

But I think that's redundant, so we can just drop it.

mcatanzaro replied:
So I've confirmed that `(closure user_data)` is redundant, but in the spirit of the "explicit is better than implicit" rule I'm planning to move them to the correct place rather than remove them. Accordingly, my suggested change is exactly as in the previous comment:

```
 * @callback: (scope async): (closure user_data): a #GAsyncReadyCallback to call when the request is satisfied
 * @user_data: the data to pass to callback function
```

I'll handle auditing all preexisting uses, so all you need to worry about are the two cases added in this pull request.

mcatanzaro replied:
I actually wound up removing these (though adding an explicit one is not harmful, as long as it's on the `callback` parameter and not the `user_data` parameter).

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitCookieManager.cpp:
```diff
@@ -517,3 +517,106 @@ void webkit_cookie_manager_delete_all_cookies(WebKitCookieManager* manager)
+    g_return_val_if_fail(WEBKIT_IS_COOKIE_MANAGER(manager), nullptr);
+    g_return_val_if_fail(g_task_is_valid(result, manager), nullptr);
+
+    return reinterpret_cast<GList*>(g_task_propagate_pointer(G_TASK(result), error));
```
Does static_cast not work here?

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitCookieManager.h.in:
```diff
@@ -79,6 +79,16 @@ typedef enum {
+ * WebKitCookieManagerOperation:
+ * @WEBKIT_COOKIE_MANAGER_REPLACE: Set new cookies and remove the existing cookies.
+ *
+ * Enum values used to denote the type of operation for setting cookies.
```
```suggestion
 * Enum values used to denote the type of operation for setting cookies.
 *
 * Since: 2.42
```

mcatanzaro reviewed (commented):
Looking good.

The API test failure is just that all API tests are broken, so adding a new test results in one newly-detected broken test. That's not your problem.

Would be good to wait for Carlos Garcia to review this too.

carlosgcampos commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -410,6 +410,26 @@ void NetworkStorageSession::setCookie(const Cookie& cookie)
+    // Delete existent cookies and add the new ones. During the process, disable
+    // signals from the cookie jar so we don't get a ton of them.
+    guint signalId = g_signal_lookup("changed", soup_cookie_jar_get_type());
+    gulong handler = g_signal_handler_find(jar, G_SIGNAL_MATCH_ID, signalId, 0, NULL, NULL, NULL);
```
```suggestion
    gulong handler = g_signal_handler_find(jar, G_SIGNAL_MATCH_ID, signalId, 0, nullptr, nullptr, nullptr);
```

carlosgcampos commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -410,6 +410,26 @@ void NetworkStorageSession::setCookie(const Cookie& cookie)
     soup_cookie_jar_add_cookie(cookieStorage(), cookie.toSoupCookie());
 }
 
+void NetworkStorageSession::setCookieJar(const Vector<Cookie>& cookies)
```
This is confusing, we have a method called `setCookieStorage` that receives a jar and this is called `setCookieJar` and receives a list of cookies. I would call this `setCookies` or `replaceCookies`

carlosgcampos commented on Source/WebCore/platform/network/NetworkStorageSession.h:
```diff
@@ -156,6 +156,9 @@ class NetworkStorageSession : public CanMakeWeakPtr<NetworkStorageSession> {
     WEBCORE_EXPORT void setCookie(const Cookie&);
     WEBCORE_EXPORT void setCookies(const Vector<Cookie>&, const URL&, const URL& mainDocumentURL);
     WEBCORE_EXPORT void setCookiesFromDOM(const URL& firstParty, const SameSiteInfo&, const URL&, std::optional<FrameIdentifier>, std::optional<PageIdentifier>, ApplyTrackingPrevention, const String&, ShouldRelaxThirdPartyCookieBlocking) const;
+#if USE(SOUP)
+    WEBCORE_EXPORT void setCookieJar(const Vector<Cookie>&);
+#endif
```
Move this to the existing `SOUP` block

carlosgcampos commented on Source/WebKit/NetworkProcess/Cookies/WebCookieManager.h:
```diff
@@ -84,6 +84,10 @@ class WebCookieManager : public NetworkProcessSupplement, public IPC::MessageRec
     void getHTTPCookieAcceptPolicy(PAL::SessionID, CompletionHandler<void(WebCore::HTTPCookieAcceptPolicy)>&&);
 
+#if USE(SOUP)
+    void setCookieJar(PAL::SessionID, const Vector<WebCore::Cookie>&, CompletionHandler<void()>&&);
```
This should be renamed too, we are not setting a new jar, but replacing the cookies in a existing jar.

carlosgcampos commented on Source/WebKit/UIProcess/API/APIHTTPCookieStore.h:
```diff
@@ -83,6 +83,9 @@ class HTTPCookieStore final : public ObjectImpl<Object::Type::HTTPCookieStore> {
 
 #if USE(SOUP)
+    void setCookieJar(Vector<WebCore::Cookie>&&, CompletionHandler<void()>&&);
+    void getCookieJar(CompletionHandler<void(const Vector<WebCore::Cookie>&)>&&);
```
getCookies or getAllCookies?

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitCookieManager.cpp:
```diff
@@ -517,3 +517,106 @@ void webkit_cookie_manager_delete_all_cookies(WebKitCookieManager* manager)
+ */
+void webkit_cookie_manager_set_all_cookies(WebKitCookieManager* manager, GList* cookies, WebKitCookieManagerOperation operation, GCancellable* cancellable, GAsyncReadyCallback callback, gpointer userData)
+{
+    g_return_if_fail(WEBKIT_IS_COOKIE_MANAGER(manager));
```
Do we allow to pass null for cookies? I guess we don't because we can `delete_all_cookies` for that.

pgorszkowski-igalia replied:
And we should not allow to pass empty list of cookies - implicit removing all cookies, right?

carlosgcampos replied:
Yes, we already have `delete_all_cookies` for that

carlosgcampos commented:
btw, I don't like the operation parameter very much, but I don't have a strong opinion, so feel free to leave it you guys think it's better. I have the feeling we will end up with a parameter that is never going to be used. I prefer to use a clearer name for the function now, like `replace_cookies` or `replace_all_cookies` and if we want to add a new one to add cookies we can add it later as `add_cookies`, `append_cookies` or whatever we want to call it.

aperezdc commented:
> btw, I don't like the operation parameter very much, but I don't have a strong opinion, so feel free to leave it you guys think it's better. I have the feeling we will end up with a parameter that is never going to be used. I prefer to use a clearer name for the function now, like `replace_cookies` or `replace_all_cookies` and if we want to add a new one to add cookies we can add it later as `add_cookies`, `append_cookies` or whatever we want to call it.

Having a different function for each operation is fine by me 👌🏼

mcatanzaro commented:
> Having a different function for each operation is fine by me 👌🏼

Sure.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5cd8455892c23c00633ea8260a3775c43ef8baa0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12639

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitCookieManager.cpp:
```diff
@@ -517,3 +517,108 @@ void webkit_cookie_manager_delete_all_cookies(WebKitCookieManager* manager)
+{
+    g_return_if_fail(WEBKIT_IS_COOKIE_MANAGER(manager));
+    g_return_if_fail(cookies);
+    g_return_if_fail(g_list_length(cookies));
```
This check is redundant, please remove it: a non-null `GList` always has elements. Quoting [the documentation](https://docs.gtk.org/glib/type_func.List.length.html):

> To check whether the list is non-empty, it is faster to check list against NULL.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitCookieManager.cpp:
```diff
@@ -517,3 +517,108 @@ void webkit_cookie_manager_delete_all_cookies(WebKitCookieManager* manager)
+
+    Vector<WebCore::Cookie> webCookies;
+    for (GList* it = cookies; it; it = g_list_next(it))
+        webCookies.append(WebCore::Cookie((SoupCookie*)it->data));
```
We can use `reinterpret_cast` here instead of a plain C cast:

```suggestion
        webCookies.append(WebCore::Cookie(reinterpret_cast<SoupCookie*>(it->data)));
```

aperezdc reviewed (approved):
I have just a couple of small comments to be fixed before landing, otherwise the patch LGTM.

Remember that you will need a second positive review as we are adding new API here.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b94d3a99d534c7dc82da1c7c264d9c94906f8827. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12639

aperezdc commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -410,6 +410,26 @@ void NetworkStorageSession::setCookie(const Cookie& cookie)
     soup_cookie_jar_add_cookie(cookieStorage(), cookie.toSoupCookie());
 }
 
+void NetworkStorageSession::replaceCookies(const Vector<Cookie>& cookies)
```
I think it's okay, the public API function has the non-empty check and that should be enough. There is no need to check again here.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitCookieManager.cpp:
```diff
@@ -517,3 +517,107 @@ void webkit_cookie_manager_delete_all_cookies(WebKitCookieManager* manager)
+ * @callback: (scope async): a #GAsyncReadyCallback to call when the request is satisfied
+ * @user_data: the data to pass to callback function
+ *
+ * Asynchronously set a list of #SoupCookie to the underlying storage.
```
```suggestion
 * Asynchronously replace all cookies in @cookie_manager with the given list of @cookies.
```

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1f2bd52c6695ae0ded759821a99c20cfab8cd578. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12639

magomez commented:
@aperezdc , @carlosgcampos , @mcatanzaro can this be merged? I think all the issues have been fixed now.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitCookieManager.cpp:
```diff
@@ -517,3 +517,106 @@ void webkit_cookie_manager_delete_all_cookies(WebKitCookieManager* manager)
+/**
+ * webkit_cookie_manager_get_all_cookies:
+ * @cookie_manager: a #WebKitCookieManager
+ * @cancellable: (allow-none): a #GCancellable or %NULL to ignore
```
```suggestion
 * @cancellable: (nullable): a #GCancellable or %NULL to ignore
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitCookieManager.cpp:
```diff
@@ -517,3 +517,106 @@ void webkit_cookie_manager_delete_all_cookies(WebKitCookieManager* manager)
+ * webkit_cookie_manager_replace_cookies:
+ * @cookie_manager: a #WebKitCookieManager
+ * @cookies: (element-type SoupCookie) (transfer none): a #GList of #SoupCookie to be added
+ * @cancellable: (allow-none): a #GCancellable or %NULL to ignore
```
```suggestion
 * @cancellable: (nullable): a #GCancellable or %NULL to ignore
```

mcatanzaro reviewed (approved):
I didn't thoroughly review this again, but I also can't think of any reason not to merge it.

Looks like it will need to be rebased for merge queue to process it. Also, there are a bunch of unresolved discussions above; don't forget to click "Resolve conversation" when you've responded to the feedback.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a21c4f44f2048c375477ae1f94e7ac2feb14d4ca. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12639

mcatanzaro commented:
Reminder: this is approved by three reviewers, but has not landed!

pgorszkowski-igalia committed:
[GTK][WPE] Add API to replace and retrieve the entire cookie jar
https://bugs.webkit.org/show_bug.cgi?id=254987

Reviewed by Carlos Garcia Campos, Adrian Perez de Castro and Michael Catanzaro.

Add possibility to replace and get all cookies, from each domain via new API.

* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
(WebCore::NetworkStorageSession::setCookieJar):
* Source/WebKit/NetworkProcess/Cookies/WebCookieManager.cpp:
(WebKit::WebCookieManager::setCookieJar):
* Source/WebKit/NetworkProcess/Cookies/WebCookieManager.h:
* Source/WebKit/NetworkProcess/Cookies/WebCookieManager.messages.in:
* Source/WebKit/UIProcess/API/APIHTTPCookieStore.cpp:
(API::HTTPCookieStore::setCookieJar):
(API::HTTPCookieStore::getCookieJar):
* Source/WebKit/UIProcess/API/APIHTTPCookieStore.h:
* Source/WebKit/UIProcess/API/glib/WebKitCookieManager.cpp:
(webkit_cookie_manager_set_cookie_jar):
(webkit_cookie_manager_set_cookie_jar_finish):
(webkit_cookie_manager_get_cookie_jar):
(webkit_cookie_manager_get_cookie_jar_finish):
* Source/WebKit/UIProcess/API/glib/WebKitCookieManager.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestCookieManager.cpp:
(testCookieManagerSetGetCookieJar):
(beforeAll):

Canonical link: https://commits.webkit.org/265834@main

webkit-commit-queue commented:
Committed 265834@main (3f2a8e4ebeb1): <https://commits.webkit.org/265834@main>

Reviewed commits have been landed. Closing PR #12639 and removing active labels.
